bolt-modules/boltlib/lib/puppet/functions/apply_prep.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+require 'fileutils'
+
+Puppet::Functions.create_function(:apply_prep) do
+  dispatch :apply_prep do
+    param 'Boltlib::TargetSpec', :targets
+  end
+
+  def apply_prep(target_spec)
+    applicator = Puppet.lookup(:apply_executor) { nil }
+    executor = Puppet.lookup(:bolt_executor) { nil }
+    inventory = Puppet.lookup(:bolt_inventory) { nil }
+    unless applicator && executor && inventory && Puppet.features.bolt?
+      raise Puppet::ParseErrorWithIssue.from_issue_and_stack(
+        Puppet::Pops::Issues::TASK_MISSING_BOLT, action: _('apply_prep')
+      )
+    end
+
+    targets = inventory.get_targets(target_spec)
+
+    # Ensure Puppet is installed
+    versions = call_function(:run_task, 'puppet_agent::version', targets)
+    need_install, installed = versions.partition { |r| r.message.chomp.empty? }
+    installed.each do |r|
+      Puppet.info "Puppet Agent #{r.message.chomp} installed on #{r.target.name}"
+    end
+    unless need_install.empty?
+      call_function(:run_task, 'puppet_agent::install', need_install.map(&:target))
+    end
+    targets.each { |target| inventory.set_feature(target, 'puppet-agent') }
+
+    # Gather facts, including custom facts
+    plugins = applicator.build_plugin_tarball do |mod|
+      search_dirs = []
+      search_dirs << mod.plugins if mod.plugins?
+      search_dirs << mod.pluginfacts if mod.pluginfacts?
+      search_dirs
+    end
+
+    task = applicator.custom_facts_task
+    results = executor.run_task(targets, task, 'plugins' => plugins)
```

Comment:
I think this needs to be a `ResultSet`. And should this function return a `ResultSet`?

Reply:
Actually, I don't understand why this isn't becoming a `ResultSet`...

Reply:
Update: it is. The local changes I was playing with were broken. :+1: